stdlib/strfromd32.c

```diff
@@ -0,0 +1,90 @@
+/* Convert _Decimal32 to string (strfromd32).
+
+   Copyright (C) 2020 Free Software Foundation, Inc.
+
+   This file is part of the Decimal Floating Point C Library.
+
+   Author(s): Joseph Kerian <jkerian@us.ibm.com>
+
+   The Decimal Floating Point C Library is free software; you can
+   redistribute it and/or modify it under the terms of the GNU Lesser
+   General Public License version 2.1.
+
+   The Decimal Floating Point C Library is distributed in the hope that
+   it will be useful, but WITHOUT ANY WARRANTY; without even the implied
+   warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
+   the GNU Lesser General Public License version 2.1 for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License version 2.1 along with the Decimal Floating Point C Library;
+   if not, write to the Free Software Foundation, Inc., 59 Temple Place,
+   Suite 330, Boston, MA 02111-1307 USA.
+
+   Please see libdfp/COPYING.txt for more information.  */
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <printf.h>
+#include <errno.h>
+#include <ctype.h>
+
+#include "printf_dfp.h"
+
+#ifndef FUNC_NAME
+#define FUNC_NAME strfromd32
+#define DECIMALN  _Decimal32
+#endif
+
+int
+FUNC_NAME (char * __restrict s, size_t n, const char * __restrict format, DECIMALN fp)
+{
+  /* format = %{[.][0-9]*[aAeEfFgG], else behavior undefined.
+     Reuse our printf hook to do this.  */
+  FILE * f = fmemopen (s, n, "w");
+  const void * const args = &fp;
+  const void * const * argsp = &args;
+  const void * const * argp = (const void * const *)&argsp;
+  struct printf_info info = { 0 };
+  int num = 0;
+
+  /* TODO: this is probably better than crashing.  The standard isn't so clear
+     about this.  */
+  if (!f)
+    {
+    /* errno set by fmemopen.  */
+    return 0;
+    }
+
+  info.user = sizeof (fp) == 4 ? mod_H : (sizeof (fp) == 8 ? mod_D : mod_DD);
```

Comment:
This may not work correctly if glibc registration fails, or a static library user does not run the onetime setup.  Possible workaround is to statically assign unique 2^n values for mod_{H,D,DD} in case registration fails or is not used.